service/matching/matcher_test.go

```diff
@@ -209,55 +209,85 @@ func (t *MatcherTestSuite) testRemoteSyncMatch(taskSource enumsspb.TaskSource) {
 	t.Equal(mustParent(t.queue.partition.(*tqid.NormalPartition), 20).RpcName(), req.GetTaskQueue().GetName())
 }
 
-func (t *MatcherTestSuite) TestRejectSyncMatchWhenBacklog() {
-	historyTask := newInternalTaskForSyncMatch(randomTaskInfo().Data, nil)
-	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+// validateSyncMatchWhenNoBacklog is a helper which validates a sync match when there is no backlog
+func (t *MatcherTestSuite) validateSyncMatchWhenNoBacklog(ctx context.Context, wg *sync.WaitGroup, historyTask *internalTask) {
+	wg.Add(1)
 
-	// sync match happens when there is no backlog
-	t.client.EXPECT().PollWorkflowTaskQueue(gomock.Any(), gomock.Any(), gomock.Any()).Return(&matchingservice.PollWorkflowTaskQueueResponse{}, errMatchingHostThrottleTest)
-	go t.childMatcher.Poll(ctx, &pollMetadata{}) //nolint:errcheck
-	time.Sleep(time.Millisecond)
-	go func() { historyTask.responseC <- nil }()
+	// poll forwarding attempt happens when there is no backlog
+	t.client.EXPECT().PollWorkflowTaskQueue(gomock.Any(), gomock.Any(), gomock.Any()).Do(
+		func(arg0 context.Context, arg1 *matchingservice.PollWorkflowTaskQueueRequest, arg2 ...interface{}) {
+			wg.Done()
+		},
+	).Return(&matchingservice.PollWorkflowTaskQueueResponse{}, errMatchingHostThrottleTest)
+
+	go func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+		_, err := t.childMatcher.Poll(ctx, &pollMetadata{})
+		t.Assert().NoError(err)
+		cancel()
+	}()
+	wg.Wait()                        // for the poll to reach the root partition
+	time.Sleep(2 * time.Millisecond) //nolint:forbidigo
+
+	go func() {
+		historyTask.responseC <- nil
+	}()
 	happened, err := t.childMatcher.Offer(ctx, historyTask)
 	t.True(happened)
 	t.Nil(err)
+}
 
+func (t *MatcherTestSuite) TestRejectSyncMatchWhenBacklog() {
+	historyTask := newInternalTaskForSyncMatch(randomTaskInfo().Data, nil)
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+
+	// sync match happens when there is no backlog
+	var wg sync.WaitGroup
+	t.validateSyncMatchWhenNoBacklog(ctx, &wg, historyTask)
+
+	// first task waits for a local poller
 	intruptC := make(chan struct{})
 	youngBacklogTask := newInternalTaskFromBacklog(randomTaskInfoWithAge(time.Second), nil)
-	t.client.EXPECT().AddWorkflowTask(gomock.Any(), gomock.Any(), gomock.Any()).Return(&matchingservice.AddWorkflowTaskResponse{}, errMatchingHostThrottleTest)
+	t.client.EXPECT().AddWorkflowTask(gomock.Any(), gomock.Any(), gomock.Any()).Do(
+		func(arg0 context.Context, arg1 *matchingservice.AddWorkflowTaskRequest, arg2 ...interface{}) {
+			wg.Done()
+		},
+	).Return(&matchingservice.AddWorkflowTaskResponse{}, errMatchingHostThrottleTest)
+	wg.Add(1)
 	go t.childMatcher.MustOffer(ctx, youngBacklogTask, intruptC) //nolint:errcheck
```

Comment:
I wonder if we should just write this test with `rootMatcher` instead of `childMatcher` so that we don't have to mock `AddWorkflowTask`. I believe this is mocked so that the tasks are not forwarded but root matcher does not have a forwarder anyway. That way the test would be much simpler.